src/wallet/scriptpubkeyman.cpp

```diff
@@ -2291,3 +2291,43 @@ bool DescriptorScriptPubKeyMan::GetDescriptorString(std::string& out, bool priv)
 
     return m_wallet_descriptor.descriptor->ToNormalizedString(provider, out, priv);
 }
+
+void DescriptorScriptPubKeyMan::UpgradeDescriptorCache()
+{
+    LOCK(cs_desc_man);
+    if (m_storage.IsLocked() || m_storage.IsWalletFlagSet(WALLET_FLAG_LAST_HARDENED_XPUB_CACHED)) {
+        return;
+    }
+
+    // Skip if we have the last hardened xpub cache
+    if (m_wallet_descriptor.cache.GetCachedLastHardenedExtPubKeys().size() > 0) {
+        return;
+    }
+
+    // Expand the descriptor
+    FlatSigningProvider provider;
+    provider.keys = GetKeys();
+    FlatSigningProvider out_keys;
+    std::vector<CScript> scripts_temp;
+    DescriptorCache temp_cache;
+    if (!m_wallet_descriptor.descriptor->Expand(0, provider, scripts_temp, out_keys, &temp_cache)){
+        throw std::runtime_error("Unable to expand descriptor");
+    }
+
+    // Cache the last hardened xpubs
+    WalletBatch batch(m_storage.GetDatabase());
+    uint256 id = GetID();
+    for (const auto& lh_xpub_pair : temp_cache.GetCachedLastHardenedExtPubKeys()) {
+        CExtPubKey xpub;
+        if (m_wallet_descriptor.cache.GetCachedLastHardenedExtPubKey(lh_xpub_pair.first, xpub)) {
```

Comment:
I don't follow.

This is checking to see if elements in the temp cache are in the stored cache;

Reply:
But the stored cache is always empty, no? I'm looking at the line 2303

Reply:
Ah, I see. There shouldn't be anything in the stored cache, however I am doing this as a belt-and-suspenders check.

Reply:
I find it confusing when we check same conditions multiple times in different places because it's not clear what is the responsibility/assumptions of each class/function. It also makes it harder to modify the code. `DescriptorScriptPubKeyMan` and `CWallet` are strongly coupled anyway.

Maybe we can just remove the check on the top of the function? Even in the worst case nothing will go wrong 1) If wallet is locked then expand will fail 2) if the cache exists already we will just regenerate it.

Reply:
This is largely just code duplicated from TopUp. The intention was to refactor them to avoid the duplication, but I never got around to doing that.

The check at the top of the function is important to avoid unnecessary extra computation which slows down wallet loading and unlocking.